File: android-client/app/src/main/java/com/example/stripeapplication/viewmodel/MainViewModel.kt
package com.example.stripeapplication.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.stripeapplication.data.models.*
import com.example.stripeapplication.data.repository.AccountRepository
import com.example.stripeapplication.data.repository.Result
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class UiState(
    val accounts: List<Account> = emptyList(),
    val selectedAccount: Account? = null,
    val paymentMethods: List<PaymentMethod> = emptyList(),
    val externalAccounts: List<ExternalAccount> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null,
    val setupIntentClientSecret: String? = null,
    val paymentIntentClientSecret: String? = null,
    val onboardingUrl: String? = null
)

class MainViewModel : ViewModel() {
    private val repository = AccountRepository()

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    init {
        loadAccounts()
    }

    fun loadAccounts() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.listAccounts()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        accounts = result.data,
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun createAccount(name: String, email: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.createAccount(name, email)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        selectedAccount = result.data,
                        successMessage = "Account created successfully",
                        isLoading = false
                    )
                    loadAccounts()
                    loadAccountData(result.data.id)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun selectAccount(account: Account) {
        _uiState.value = _uiState.value.copy(selectedAccount = account)
        loadAccountData(account.id)
    }

    fun loadAccountData(accountId: String) {
        viewModelScope.launch {
            // Load account details
            when (val result = repository.getAccount(accountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(selectedAccount = result.data)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(error = result.message)
                }
            }
        }
        loadPaymentMethods(accountId)
        loadExternalAccounts(accountId)
    }

    fun deleteAccount(accountId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.deleteAccount(accountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        selectedAccount = null,
                        paymentMethods = emptyList(),
                        externalAccounts = emptyList(),
                        successMessage = "Account deleted successfully",
                        isLoading = false
                    )
                    loadAccounts()
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun upgradeToRecipient(accountId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.upgradeToRecipient(accountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Account upgraded to recipient",
                        isLoading = false
                    )
                    loadAccountData(accountId)
                    loadAccounts()
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun createOnboardingLink(accountId: String, refreshUrl: String, returnUrl: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.createOnboardingLink(accountId, refreshUrl, returnUrl)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        onboardingUrl = result.data.url,
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun clearOnboardingUrl() {
        _uiState.value = _uiState.value.copy(onboardingUrl = null)
    }

    // Payment Methods
    fun loadPaymentMethods(accountId: String) {
        viewModelScope.launch {
            when (val result = repository.listPaymentMethods(accountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(paymentMethods = result.data)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(error = result.message)
                }
            }
        }
    }

    fun createSetupIntent(accountId: String, customerId: String? = null) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.createSetupIntent(accountId, customerId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        setupIntentClientSecret = result.data.clientSecret,
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun clearSetupIntent() {
        _uiState.value = _uiState.value.copy(setupIntentClientSecret = null)
    }

    fun onSetupIntentConfirmed(accountId: String) {
        _uiState.value = _uiState.value.copy(
            setupIntentClientSecret = null,
            successMessage = "Card added successfully"
        )
        loadPaymentMethods(accountId)
    }

    fun deletePaymentMethod(accountId: String, paymentMethodId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.deletePaymentMethod(accountId, paymentMethodId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Payment method removed",
                        isLoading = false
                    )
                    loadPaymentMethods(accountId)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    // External Accounts
    fun loadExternalAccounts(accountId: String) {
        viewModelScope.launch {
            when (val result = repository.listExternalAccounts(accountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(externalAccounts = result.data)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(error = result.message)
                }
            }
        }
    }

    fun createExternalAccount(accountId: String, token: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.createExternalAccount(accountId, token)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Bank account added successfully",
                        isLoading = false
                    )
                    loadExternalAccounts(accountId)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun deleteExternalAccount(accountId: String, externalAccountId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.deleteExternalAccount(accountId, externalAccountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Bank account removed",
                        isLoading = false
                    )
                    loadExternalAccounts(accountId)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun setDefaultExternalAccount(accountId: String, externalAccountId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.setDefaultExternalAccount(accountId, externalAccountId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Default bank account updated",
                        isLoading = false
                    )
                    loadExternalAccounts(accountId)
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    // Transactions
    fun payUser(
        accountId: String,
        recipientAccountId: String,
        paymentMethodId: String,
        amount: Int
    ) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.payUser(accountId, recipientAccountId, paymentMethodId, amount)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        successMessage = "Payment of $${amount / 100.0} sent successfully!",
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun createPaymentIntent(
        accountId: String,
        recipientAccountId: String,
        amount: Int,
        savePaymentMethod: Boolean
    ) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            when (val result = repository.createPaymentIntent(accountId, recipientAccountId, amount, savePaymentMethod)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        paymentIntentClientSecret = result.data.clientSecret,
                        isLoading = false
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        error = result.message,
                        isLoading = false
                    )
                }
            }
        }
    }

    fun clearPaymentIntent() {
        _uiState.value = _uiState.value.copy(paymentIntentClientSecret = null)
    }

    fun onPaymentIntentConfirmed(accountId: String) {
        _uiState.value = _uiState.value.copy(
            paymentIntentClientSecret = null,
            successMessage = "Payment completed successfully!"
        )
        loadPaymentMethods(accountId)
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    fun clearSuccessMessage() {
        _uiState.value = _uiState.value.copy(successMessage = null)
    }
}

File: ios-client/StripeApplication/StripeApplication/MainViewModel.swift
import Foundation
import SwiftUI
import Observation

@Observable
class MainViewModel {
    var accounts: [Account] = []
    var selectedAccount: Account? = nil
    var paymentMethods: [PaymentMethod] = []
    var externalAccounts: [ExternalAccount] = []
    var isLoading: Bool = false
    var error: String? = nil
    var successMessage: String? = nil
    var setupIntentClientSecret: String? = nil
    var paymentIntentClientSecret: String? = nil
    var onboardingUrl: String? = nil

    private let api = APIClient.shared

    init() {
        Task { await loadAccounts() }
    }

    // MARK: - Accounts

    func loadAccounts() async {
        isLoading = true
        error = nil
        do {
            accounts = try await api.listAccounts()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func createAccount(name: String, email: String) async {
        isLoading = true
        error = nil
        do {
            let account = try await api.createAccount(name: name, email: email)
            selectedAccount = account
            successMessage = "Account created successfully"
            isLoading = false
            await loadAccounts()
            await loadAccountData(accountId: account.id)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func selectAccount(_ account: Account) {
        selectedAccount = account
        Task { await loadAccountData(accountId: account.id) }
    }

    func loadAccountData(accountId: String) async {
        do {
            let account = try await api.getAccount(accountId: accountId)
            selectedAccount = account
        } catch {
            self.error = error.localizedDescription
        }
        await loadPaymentMethods(accountId: accountId)
        await loadExternalAccounts(accountId: accountId)
    }

    func deleteAccount(accountId: String) async {
        isLoading = true
        error = nil
        do {
            try await api.deleteAccount(accountId: accountId)
            selectedAccount = nil
            paymentMethods = []
            externalAccounts = []
            successMessage = "Account deleted successfully"
            isLoading = false
            await loadAccounts()
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func upgradeToRecipient(accountId: String) async {
        isLoading = true
        error = nil
        do {
            _ = try await api.upgradeToRecipient(accountId: accountId)
            successMessage = "Account upgraded to recipient"
            isLoading = false
            await loadAccountData(accountId: accountId)
            await loadAccounts()
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func createOnboardingLink(accountId: String) async {
        isLoading = true
        error = nil
        do {
            let response = try await api.createOnboardingLink(
                accountId: accountId,
                refreshUrl: "https://example.com/refresh",
                returnUrl: "https://example.com/return"
            )
            onboardingUrl = response.url
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func clearOnboardingUrl() {
        onboardingUrl = nil
    }

    // MARK: - Payment Methods

    func loadPaymentMethods(accountId: String) async {
        do {
            paymentMethods = try await api.listPaymentMethods(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func createSetupIntent(accountId: String, customerId: String? = nil) async {
        isLoading = true
        error = nil
        do {
            let response = try await api.createSetupIntent(accountId: accountId, customerId: customerId)
            setupIntentClientSecret = response.clientSecret
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func clearSetupIntent() {
        setupIntentClientSecret = nil
    }

    func onSetupIntentConfirmed(accountId: String) {
        setupIntentClientSecret = nil
        successMessage = "Card added successfully"
        Task { await loadPaymentMethods(accountId: accountId) }
    }

    func deletePaymentMethod(accountId: String, paymentMethodId: String) async {
        isLoading = true
        error = nil
        do {
            try await api.deletePaymentMethod(accountId: accountId, paymentMethodId: paymentMethodId)
            successMessage = "Payment method removed"
            isLoading = false
            await loadPaymentMethods(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    // MARK: - External Accounts

    func loadExternalAccounts(accountId: String) async {
        do {
            externalAccounts = try await api.listExternalAccounts(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func createExternalAccount(accountId: String, token: String) async {
        isLoading = true
        error = nil
        do {
            _ = try await api.createExternalAccount(accountId: accountId, token: token)
            successMessage = "Bank account added successfully"
            isLoading = false
            await loadExternalAccounts(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func deleteExternalAccount(accountId: String, externalAccountId: String) async {
        isLoading = true
        error = nil
        do {
            try await api.deleteExternalAccount(accountId: accountId, externalAccountId: externalAccountId)
            successMessage = "Bank account removed"
            isLoading = false
            await loadExternalAccounts(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func setDefaultExternalAccount(accountId: String, externalAccountId: String) async {
        isLoading = true
        error = nil
        do {
            _ = try await api.setDefaultExternalAccount(accountId: accountId, externalAccountId: externalAccountId)
            successMessage = "Default bank account updated"
            isLoading = false
            await loadExternalAccounts(accountId: accountId)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func createBankAccountToken(
        accountId: String,
        accountHolderName: String,
        routingNumber: String,
        accountNumber: String
    ) async {
        isLoading = true
        error = nil
        do {
            let tokenId = try await api.createBankAccountToken(
                accountHolderName: accountHolderName,
                routingNumber: routingNumber,
                accountNumber: accountNumber
            )
            await createExternalAccount(accountId: accountId, token: tokenId)
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    // MARK: - Transactions

    func payUser(accountId: String, recipientAccountId: String, paymentMethodId: String, amount: Int) async {
        isLoading = true
        error = nil
        do {
            _ = try await api.payUser(
                accountId: accountId,
                recipientAccountId: recipientAccountId,
                paymentMethodId: paymentMethodId,
                amount: amount
            )
            successMessage = "Payment of $\(String(format: "%.2f", Double(amount) / 100.0)) sent successfully!"
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func createPaymentIntent(accountId: String, recipientAccountId: String, amount: Int, savePaymentMethod: Bool) async {
        isLoading = true
        error = nil
        do {
            let response = try await api.createPaymentIntent(
                accountId: accountId,
                recipientAccountId: recipientAccountId,
                amount: amount,
                savePaymentMethod: savePaymentMethod
            )
            paymentIntentClientSecret = response.clientSecret
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func clearPaymentIntent() {
        paymentIntentClientSecret = nil
    }

    func onPaymentIntentConfirmed(accountId: String) {
        paymentIntentClientSecret = nil
        successMessage = "Payment completed successfully!"
        Task { await loadPaymentMethods(accountId: accountId) }
    }

    func clearError() {
        error = nil
    }

    func clearSuccessMessage() {
        successMessage = nil
    }
}
